### PR TITLE
LPD-51759 - FDS + Info Panel

### PR DESCRIPTION
**Pull request by juanjofgliferay**
## Story / Task
[LPD-43470](https://liferay.atlassian.net/browse/LPD-43470) / [LPD-51759](https://liferay.atlassian.net/browse/LPD-51759)

> [!WARNING]  
> Feature under Feature Flag. 
> `feature.flag.LPD-41774=true`

## Goal
Add support to handle Info Panel from the FDS component allowing consumers to provide their own component

## Concerns
Actual code relies on a local copy of the [Clay Side Panel ](https://github.com/liferay/clay/pull/6058) that it is still WIP

## UI (FDS Sample)

https://github.com/user-attachments/assets/1d024388-4237-4381-8152-96e2eb4ed40b

## Old UI (CMS) 

Screenshot from 2025-05-23 09-18-21

### CODE REVIEW

File: modules/apps/frontend-data-set/frontend-data-set-web/package.json [hunk 1]
```diff
@@ -25,6 +25,7 @@
 		"frontend-js-components-web": "*",
 		"frontend-js-web": "*",
 		"jest-fetch-mock": "3.0.3",
+		"motion": "^12.12.1",
```
Comment (juanjofgliferay) [inline modules/apps/frontend-data-set/frontend-data-set-web/package.json:28]: This is needed as part of the Clay component implementation

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/info_panel/InfoPanel.tsx [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import React, {useContext} from 'react';
+
+import FrontendDataSetContext from '../FrontendDataSetContext';
+import {SidePanel, SidePanelProps} from './clay_side_panel';
+
+export function InfoPanel({component: InfoPanelContent, ...props}: any) {
+	const {selectedItems} = useContext(FrontendDataSetContext);
+
+	if (InfoPanelContent) {
+		return (
```
Comment (juanjofgliferay) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/info_panel/InfoPanel.tsx:16]: We can be more opinionated about the information we want to pass to the Info Panel. Right now, we are passing all selected items so the responsability of displaying the counter of items selected or the content relies on the developer.
We can pass `selectedItems` **only** if one is selected and the `selectedItemsLength` otherwise.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/info_panel/InfoPanel.tsx [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+			<SidePanel {...props}>
+				<InfoPanelContent items={selectedItems} />
+			</SidePanel>
+		);
+	}
+
+	return (
+		<SidePanel {...props}>
```
Comment (dsanz) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/info_panel/InfoPanel.tsx:24]: this varian't won't kick in as long as FDS client code passes an info panel content in, so do we want to keep it?

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.tsx [hunk 4]
```diff
@@ -1007,7 +1137,10 @@ const FrontendDataSet = ({
 			}}
 		>
 			<ViewsContext.Provider value={[viewsState, viewsDispatch]}>
-				<div className="fds">
+				<div
+					className="fds"
+					style={{paddingRight: isInfoPanelOpen ? 320 : 0}}
```
Comment (juanjofgliferay) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.tsx:1142]: There is no smooth animation yet. Waiting for the final Clay component implementation or pattern to apply the correct transition here.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/info_panel/InfoPanel.tsx (hunk 3 above)
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/info_panel/InfoPanel.tsx:24]: It is just a safe check in case somebody accidentally set the `showInfoPanel: true` but do not provide the template. Can be removed :broom:

**Comment by juanjofgliferay** [comment]
ci:test:sf

File: modules/apps/frontend-data-set/frontend-data-set-web/package.json (hunk 1 above)
Comment (dsanz) [reply modules/apps/frontend-data-set/frontend-data-set-web/package.json:28]: Once we import component from clay, this will be gone won't it?

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/info_panel/InfoPanel.tsx (hunk 3 above)
Comment (dsanz) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/info_panel/InfoPanel.tsx:24]: I'm fine with providing a default content for the info panel. 

We have to think about the `showInfoPanel` flag. Is it necessary? If all FDS instances requiring an info panel must pass an infopanel component externally (or just an URL), the component not being `undefined` is the flag we're looking for.

We can enable the info panel only if such component is provided.

Am I missing some corner case where we want to turn on the info panel with no external component? maybe if we provide a default one for testing?

**Comment by juanjofgliferay** [comment]
ci:test:frontend

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.tsx [hunk 5]
```diff
@@ -120,16 +122,21 @@ const FrontendDataSet = ({
...
 	views,
 }: IFrontendDataSetProps) => {
+	const fdsRef = useRef(null);
 	const wrapperRef = useRef(null);
 	const [componentLoading, setComponentLoading] = useState(false);
 	const [creationMenu, setCreationMenu] = useState(initialCreationMenu);
 	const [dataLoading, setDataLoading] = useState(!!apiURL);
+	const [dataSetSupportInfoPanelId] = useState(
+		`support-info-panel-${getRandomId()}`
+	);
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.tsx:132]: This is technically not wrong, but the typical use of `useState` is to leverage re-render after state change. For a value that never changes and does not cause any re-render, typical recommend is `useRef`.

**Review by markocikos — COMMENTED** [review]
@juanjofgliferay few comments and questions inline

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.tsx (hunk 5 above)
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.tsx:132]: Just following the same pattern used for similar items

```
const [dataSetSupportModalId] = useState(`support-modal-${getRandomId()}`);
const [dataSetSupportSidePanelId] = useState(
	sidePanelId || `support-side-panel-${getRandomId()}`
);
```

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.ts [hunk 6]
```diff
@@ -24,6 +24,7 @@ export declare function FrontendDataSet({
 	formName,
 	header,
 	id,
+	infoPanel,
```
[the remark this answers is not in the document]
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.ts:27]: It seems that it is spread across the portal :sweat_smile: 

image

or
image

[the remark this answers is not in the document]
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.ts:27]: I'd rather prefer doing in a follow up :sweat:

**Comment by juanjofgliferay** [comment]
ci:test:sf

**Comment by juanjofgliferay** [comment]
@markocikos Added one last commit that includes topic of [LPD-49157 story](https://liferay.atlassian.net/browse/LPD-49157) (open info panel from different triggers)

**Comment by juanjofgliferay** [comment]
ci:test:relevant

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.tsx [hunk 7]
```diff
@@ -877,6 +883,10 @@ const FrontendDataSet = ({
 			});
 	}
 
+	function openInfoPanel() {
+		setIsInfoPanelOpen((value) => !value);
+	}
```
[the remark this answers is not in the document]
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.tsx]: In `InfoPanelToggleButton`, you can have something like:

```js
return (
    <ClayButtonWithIcon
        onClick={onInfoPanelToggleButtonClick}
    />
);
```
where `onInfoPanelToggleButtonClick` is a prop in `FrontendDataSetContext`.

Then, in `FrontendDataSet`, you can have 

```js
<FrontendDataSetContext.Provider
    value={{
        onItemsChange,
        onInfoPanelToggleButtonClick: () => {
            setInfoPanelOpen((value) => !value);
        }
        onSearch,
    }}
>
```
It's a subtle difference, but important one. We are invoking state change in the component holding that state. Inner component is not aware of that state, it does not control it. It just has a generic click handler.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.tsx (hunk 5 above)
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.tsx]: see https://github.com/liferay-frontend/liferay-portal/pull/4924#discussion_r2142986988

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.tsx]: I don't mind inline the logic as you suggest, but I think that generic `openInfoPanel` is more clear, as it is also used by item actions of type `infoPanel`, quick actions and the `ActionLinkRenderer`

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/index.ts]: As we already have an `index.ts`, I tried moving `index.js` to `index.ts` but ending up with an unrelated error in `objects-web` :shrug: after running `yarn node-scripts check:tsc --all` to ensure that there were no side effects.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.tsx]: Yep, makes sense.

**Comment by juanjofgliferay** [comment]
ci:forward:force
